Improve page conversion attribution performance with pre-calculated field

bx80 opened the pull request:
### Description:

Fixes #20375

- Adds a migration to create a new `log_conversion.pageviews_before` column (> 4.14.0 installs will already have this).
- New conversions will now write the number of pageviews in the visit prior to the conversion into this field.
- Adds a new console command `core:calculate-conversion-pages` to populate the new field for historic data.
- The migration will automatically attempt to populate the new field for conversions the last 48hrs, providing the host is not *.matomo.cloud and there are less than 10,000 conversions in that period.
- The pages / goals archiving query has been adjusted to use the new field instead of an expensive sub query.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
Migration to create new pageviews_before field on log_conversion, populate field on insert conversion, console command to calculate history

Commit:
Migration to create new pageviews_before field on log_conversion, populate field on insert conversion, console command to calculate history

Commit:
Migration to create new pageviews_before field on log_conversion, populate field on insert conversion, console command to calculate history

Commit:
Migration to create new pageviews_before field on log_conversion, populate field on insert conversion, console command to calculate history

Commit:
Migration to create new pageviews_before field on log_conversion, populate field on insert conversion, console command to calculate history

Commit:
Migration to create new pageviews_before field on log_conversion, populate field on insert conversion, console command to calculate history

sgiehl commented on core/Tracker/Model.php:
```diff
@@ -35,8 +43,19 @@ public function createAction($visitAction)
     public function createConversion($conversion)
     {
+        // Calculate and store the pages viewed before this conversion
+        $conversion['pageviews_before'] = $this->getPageCountViewedBeforeDate($conversion['idvisit'], $conversion['server_time']);
```
Any reason for not handling this and also the database column stuff using a ConversionDimension class like https://github.com/matomo-org/matomo/blob/5.x-dev/plugins/ExampleTracker/Columns/ExampleConversionDimension.php#L27

bx80 replied:
A couple of reasons:

- The `onGoalConversion` event doesn't provide access to the conversion server_time field needed for the query. Using the current time could theoretically cause inaccuracy on a slow server if the conversion server time was say 23:00:00.999, an additional (parallel) action was performed at 23:00:01.200 and the `onGoalConversion` runs at 23:00:01.400, then an additional pageview would be incorrectly counted. 

- ConversionDimension classes don't usually perform additional queries and the tracker db object is not exposed for the event. It would be more efficient to perform the query using the existing tracker db connection.

The performance and accuracy gains might well be negligible, and it would be good to be consistent. So I'm happy to rework this to use a dimension class if you think that's a better approach in this case?

sgiehl replied:
I'm actually unsure if having an additional query being performed each time a conversion is triggered might not be a bit too heavy. Especially as this one even uses a join. Any query performed during tracking needs to be super fast.

I guess this could be solved without having the additional query at all if #17466 would be merged before. This one adds an additional dimension to the log_visit table, which counts the pageviews only of the current visit. Your query currently looks up all actions that were performed in the visit before the conversion. That new dimension would actually always hold that value at the time the conversion occurs.

@tsteur would you mind having a quick look here? I'm not sure if adding a new query performed while tracking conversions is that much of a benefit compared to having a slower archiving. I would favor to solve that using #17466, but that was postponed to Matomo 6, due to the additional visit dimension.

bx80 replied:
Great idea @sgiehl, #17466 seems like an excellent way to avoid the extra query and a useful addition on it's own :+1:

tsteur replied:
Indeed we likely wouldn't want that query as the tracking part would cause more load then the query. Like some other fields can we maybe copy over the `pageview_position` from the log_link_visit_action? We may have this information already in some cases already?
Potentially another way might be `log_visit.visit_total_interactions - log_visit.visit_total_searches`?

sgiehl replied:
All other fields also contain events or other type of actions. But we actually only want to know the page views. The other PR would add a dimension for exactly that purpose, but would mean a longer running migration though.

tsteur replied:
Just to confirm: `log_visit.visit_total_interactions - log_visit.visit_total_searches` should give us the amount of page views? Or are you saying it includes events? From what I see it should be the same as what is implemented in #17466 without needing a migration?

bx80 replied:
https://github.com/matomo-org/matomo/blob/aea8f07de8ea3ce6ce39d3a217af5ec6a957db96/plugins/Actions/Columns/VisitTotalInteractions.php#L92

From what I can see the `VisitTotalInteractions` metric only counts actions if they have a URL or are `Action::TYPE_SITE_SEARCH` so `log_visit.visit_total_interactions - log_visit.visit_total_searches` looks like it will work. :slightly_smiling_face: 

Only possible issue is that that visitInfo values to be incremented get overwritten by other request processors so the `visitTotalInteractions` vlaue will be `'visit_total_interactions + 1'` instead of the original numeric value. I can add an option to the `VisitProperties` class to store a read only copy of the original visitInfo values to work around this.

tsteur commented on core/Updates/5.0.0-b1.php:
```diff
@@ -110,4 +117,39 @@ private function requiresUpdatedLogVisitTableIndex()
+    {
+        // Abort if host is *.matomo.cloud
+        $piwikUrl = SettingsPiwik::getPiwikUrl();
+        if (strpos($piwikUrl, '.matomo.cloud') !== false) {
```
fyi there could be also `.innocraft.cloud`. And we would need to use `SettingsPiwik::getPiwikInstanceId()` as there could be various custom domains on it and then it wouldn't be detected.

Commit:
Use a ConversionDimension class instead of directly modifying the conversion insert, calculate value from other fields instead of using a query, VisitInfo to provide access to original visit values.

Commit:
Use a ConversionDimension class instead of directly modifying the conversion insert, calculate value from other fields instead of using a query, VisitInfo to provide access to original visit values.

Commit:
Use a ConversionDimension class instead of directly modifying the conversion insert, calculate value from other fields instead of using a query, VisitInfo to provide access to original visit values.

Commit:
Use a ConversionDimension class instead of directly modifying the conversion insert, calculate value from other fields instead of using a query, VisitInfo to provide access to original visit values.

Commit:
Added tests

Commit:
Added tests

Commit:
Added tests

Commit:
Added tests

Commit:
Added tests

Commit:
Added tests

Commit:
Fix missing parameters for ecommerce conversions

Commit:
Fix missing parameters for ecommerce conversions

Commit:
Fix test

Commit:
Fix test

Commit:
Fix test

Commit:
Fix test

Commit:
Fix for UI test

Commit:
Fix for UI test

Commit:
Fix for UI test

Commit:
Fix for UI test

Commit:
Update UI test screenshots, update privacy manager tests, update dimension test

Commit:
Update UI test screenshots, update privacy manager tests, update dimension test

Commit:
Update UI test screenshots, update privacy manager tests, update dimension test

Commit:
Update UI test screenshots, update privacy manager tests, update dimension test

Commit:
Fix query

Commit:
Fix query

Commit:
Fix query

Commit:
Fix query

Commit:
Fix query

Commit:
Fix query

Commit:
Fix archiving query segmentation support, fix update query, adjust aggregation of new pages before metric

Commit:
Fix archiving query segmentation support, fix update query, adjust aggregation of new pages before metric

Commit:
Fix archiving query segmentation support, fix update query, adjust aggregation of new pages before metric

Commit:
Fix archiving query segmentation support, fix update query, adjust aggregation of new pages before metric

Commit:
Fix archiving query segmentation support, fix update query, adjust aggregation of new pages before metric

Commit:
Fix archiving query segmentation support, fix update query, adjust aggregation of new pages before metric

Commit:
Fix for page before metric aggregation

Commit:
Fix for page before metric aggregation

Commit:
Fix for page before metric aggregation

Commit:
Fix for page before metric aggregation

Commit:
Fix for page before metric aggregation

Commit:
Fix for page before metric aggregation

sgiehl commented on plugins/Goals/Columns/PageviewsBefore.php:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @return mixed|false
+     */
+    public function onEcommerceCartUpdateConversion(Request $request, Visitor $visitor, $action, GoalManager $goalManager)
```
Isn't that one triggered when an ecommerce cart is updated? As that one is not a real "conversion" and won't appear in any conversion reports I don't think we should set the value for it.

bx80 replied:
Good spot, removed :+1:

sgiehl commented on core/Tracker/Visitor.php:
```diff
@@ -42,10 +42,10 @@ public function setVisitorColumn($column, $value)
     }
 
-    public function getVisitorColumn($column)
+    public function getVisitorColumn($column, bool $original = false)
```
Those changes are actually a part from what #19316 was supposed to "fix".
I'm actually not sure if adding additional parameters to the methods is the best solution. It for sure works, but we should maybe prefer a solution that makes it very clear with which values you are working with based the methods name.

Instead of changing the VisitProperties class to hold original and updated values, we should simply keep the class as it is and hold an additional instance of it. Maybe the class could have an additional constructor parameter to define if the properties are manipulate-able or not. So we can define the original values and the one from previous visit as final and trying to use a method to change the properties would then throw an exception.

Also with #19316 we should consider reworking methods like get `getPreviousVisitColumn` and  maybe introduce a new set of methods that make it clear which values will be used. To me `getPreviousVisitColumn` didn't make it clear if it contains the original values from the current visits or values from a visit before the current one. But that could maybe simply be solve with useful docblocks.

We can maybe also postpone further work on this issue and first solve #19316

bx80 replied:
While an additional method parameter to return the original visitor property value might not be the perfect solution, it requires a lot less rework and breaking changes than introducing another `VisitorProperty` object which must then be passed around by the methods in` VisitorRecognizer` and the many request processors.

Given this does work and solves the immediate requirement, perhaps we can consider it a short term solution to be replaced by a more comprehensive solution if/when #19316 is implemented?

bx80 replied:
I've created a [PR](https://github.com/matomo-org/matomo/pull/20600) to fix #19316 and rebased this PR on it.

sgiehl commented on core/Updates/5.0.0-b1.php:
```diff
@@ -110,4 +117,39 @@ private function requiresUpdatedLogVisitTableIndex()
+     */
+    private function getPagesBeforeAddColumn()
+    {
+        return $this->migration->db->addColumns('log_conversion', ['pageviews_before' => "SMALLINT UNSIGNED DEFAULT NULL"]);
```
Is that one needed? I thought this would be done by the dimension class automatically.

bx80 replied:
This was added prior to deciding to use a  dimension class - now removed :+1:

sgiehl commented on core/Updates/5.0.0-b1.php:
```diff
@@ -110,4 +117,39 @@ private function requiresUpdatedLogVisitTableIndex()
+    {
+        // Abort if host is *.matomo.cloud or *.innocraft.cloud
+        $instanceId = SettingsPiwik::getPiwikInstanceId();
+        if (strpos($instanceId, '.matomo.cloud') !== false || strpos($instanceId, '.innocraft.cloud') !== false) {
```
I don't think we should add such checks in core. If cloud doesn't want to apply certain updates, they might need to adjust that themselves. Also I guess that check wouldn't work for customers with custom domains.

@tsteur Is there any process for updates that should not be done on cloud automatically?

tsteur replied:
FYI it would actually work for custom domains as we're checking the instanceId and not the matomo url.

There is no process yet for this AFAIK. There could be an issue creation but this may result in a bit of manual work. Or maybe if it's only used in update script, maybe it wouldn't be too bad to have a `isCloud` method and then not execute it.

bx80 replied:
So is it okay to leave this check as is since it's just a update script?  I could adjust it to check if the cloud plugin is loaded instead if that would be more reliable?

sgiehl replied:
But don't we need to run the update on cloud as well? Otherwise the calculated numbers for today might not be correct afterwards... Or is the purpose to be able to run the update manually and not all at once?

tsteur replied:
We may handle that manually @sgiehl or ideally, if there are say less than 500 conversions in the last 24 hours or simply generally we could always reprocess the last 500 conversions. This would solve it for many. We should check the performance of the query though ideally as @bx80 suggested

sgiehl replied:
I had another thought on that one. I still think we should remove the check. If we think the query is not fast enough to run the migration automatically on cloud, we should not include the migration at all. There are on premise installs out there that are a lot bigger than our biggest cloud instances and we shouldn't simply let those instances run into problems we are aware of.

So we could try running the updating query on a bigger dataset/instance upfront to check how much data we could process automatically and implement that limit (if 10.000 is too high).
Or alternatively we lower that limits to e.g. 500, as suggested by @tsteur, and somehow advise on premise users that they might need to run the command

bx80 replied:
@sgiehl I did some testing of the query speed and it's quite fast, 10,000 conversions updated in under 10 seconds and 1,000,000 in under a minute. See my comment [here](https://github.com/matomo-org/matomo/pull/20526#discussion_r1175921101). Since we're only processing the last 48hrs then I think a 10,000 limit should be fast enough for even large on premise users? 

With cloud we'd be running the update on thousands of customers at the same time using the same shared resources, so even if any individual database takes ~30 seconds that could be ~80 hours per 10,000

Would it make sense that if we detect more than 10,000 conversions in the last 48hrs during the upgrade and do not perform the update query then we set an option that can then be used to show a diagnostic warning that shows  the command to run?

bx80 replied:
After discussion in the sprint review it was decided to remove the check since the update query is fast and executed separately per goal/site combination. Cloud will not run the update query at all.

sgiehl commented on core/Updates/5.0.0-b1.php:
```diff
@@ -110,4 +117,39 @@ private function requiresUpdatedLogVisitTableIndex()
+        $sql = 'SELECT COUNT(*) FROM ' . Common::prefixTable('log_conversion') . ' WHERE server_time <= ? AND server_time >= ?';
+
+        $result = Db::fetchOne($sql, [$startDate, $endDate]);
+        if ($result > 10000) {
```
I guess this check is to not "harm" bigger installs with long running queries. But as the user will never see if the update ran or not it could end up in incorrect data if the command to update the values wasn't used.

bx80 replied:
Since all the new values are calculated for each goal using a single query perhaps the query performance should be tested on a large dataset to get some idea of performance? If it's able to do ~1,000,000 conversions in a reasonable time then perhaps there is no need for a limit or the limit could be increased?

```SQL
UPDATE log_conversion c
LEFT JOIN (
     SELECT COUNT(va.idvisit) AS pages_before, va.idvisit, va.server_time
     FROM log_link_visit_action va
     LEFT JOIN log_action a ON a.idaction = va.idaction_url AND a.type = 1
     GROUP BY va.idvisit
) AS a ON a.idvisit = c.idvisit AND a.server_time <= c.server_time
SET c.pageviews_before = a.pages_before
WHERE c.idsite = ? AND c.idgoal = ? AND c.pageviews_before IS NULL;
```

bx80 replied:
@tsteur To test this I generated a test dataset locally with 3,000,000 conversion records, each with 20 linked visit actions before conversion. Then I ran the update query progressively decreasing the dataset size. 

| Conversions | Query time |
| ---: | ---:|
|3,000,000 |181.2s |
|1,000,000 | 58.9s |
|300,000  |24.1s |
|100,000 | 9.6s. |
|10,000 |1.6s |

That seems fairly quick. 

How long do we consider a reasonable time to be during an update? 
We could also a separate limit lower for cloud and a higher limit (or perhaps none?) for on- premise?

tsteur replied:
@bx80 On Cloud we would execute this update query after the deploy of Matomo 5 for it to process everything else quickly. Meaning we would ideally have a custom command that we could execute that only executes say the most recent X conversions (say the most recent 50K conversions).

bx80 replied:
@tsteur The `core:calculate-conversion-pages` command included in this PR has a date range option, but I can add another option to process just the last X conversions instead. :+1: 

So to summarize, the migration script will not attempt to update any existing conversions for cloud domains, and the new command to update the last X conversions can be run separately after deployment by the cloud team.

tsteur replied:
Awesome, that sounds great @bx80

sgiehl reviewed (changes requested):
Left a couple of comments. Otherwise I was wondering if the current approach of simply replacing the archiving logic with a one based on the new column together with providing a command to recalculate the columns is suitable for us.

An alternative would be to keep the old possibility to calculate the values and decide whether to use the new or the old method after checking if the new column has values for all dates within the requested date range or not.
This would remove the requirements of having a command to update old datasets.

Using the current approach would otherwise mean that we would need to run the command for all cloud installs even if archives are already built. Otherwise we would have incorrect numbers after invalidating certain archives or if e.g. a new segment or custom report is created that triggers archiving for date ranges before the update.

@tsteur Which solution would you prefer?

tsteur commented:
> An alternative would be to keep the old possibility to calculate the values and decide whether to use the new or the old method after checking if the new column has values for all dates within the requested date range or not.
This would remove the requirements of having a command to update old datasets.

> Using the current approach would otherwise mean that we would need to run the command for all cloud installs even if archives are already built. Otherwise we would have incorrect numbers after invalidating certain archives or if e.g. a new segment or custom report is created that triggers archiving for date ranges before the update.

On the cloud it shouldn't cause any issues unless someone invalidated historical data and this should be rare and is ok if then the data wouldn't show up correctly. All other data would be already archived using the old logic and the new logic would then start for newly tracked data. We may just have to populate the data for the last 24 hours to ensure the current day is correct. It would also apply to new segments/custom reports but we're only archiving 1 month back on Cloud so this wouldn't be too concerning and the report is potentially also not viewed too often. In the worst case we'd rather invalidate the data and re-archive it if someone asks.

I'm thinking checking all the column values be quite time consuming on the database and may take quite a bit of time. It would only apply for Cloud for invalidated data so it wouldn't be needed. On the Cloud we'd rather not have that extra query.

Commit:
Remove onCartUpdate calculation, remove unnecessary column creation.

Commit:
Remove onCartUpdate calculation, remove unnecessary column creation.

Commit:
Remove onCartUpdate calculation, remove unnecessary column creation.

Commit:
Remove onCartUpdate calculation, remove unnecessary column creation.

Commit:
Update system test result

Commit:
Update system test result

Commit:
Update system test result

Commit:
Update system test result

Commit:
Add immutable visitor property functions to the VisitProperties class, Visitor class and populate via the VisitRecognizer. Adjust referrer attribution and tests to use the immutable properties.

Commit:
Add immutable visitor property functions to the VisitProperties class, Visitor class and populate via the VisitRecognizer. Adjust referrer attribution and tests to use the immutable properties.

Commit:
Update to use getImmutableVisitorColumn() method

Commit:
Update to use getImmutableVisitorColumn() method

Commit:
Only use immutable properties for the isCurrentReferrerDirectEntry() check

Commit:
Only use immutable properties for the isCurrentReferrerDirectEntry() check

Commit:
Fix referrer tests

Commit:
Fix referrer tests

Commit:
Added last-n option to core:calculate-conversion-pages

Commit:
Added last-n option to core:calculate-conversion-pages

Commit:
Update command for vendor proxy changes

Commit:
Update command for vendor proxy changes

bx80 commented:
Because we can't do `UPDATE ... LEFT JOIN ... LIMIT` in MySQL I've implemented the `--last-n` command option by querying the `server_time` value of the last-n conversion and then using it to set a date range for the update query using the where clause. 

The command has also been reworked to take into account the vendor proxy changes and rebased onto 5.x-dev now that the [immutable properties](https://github.com/matomo-org/matomo/pull/20600) change has been merged.

Commit:
Set parameter default value

Commit:
Set parameter default value

sgiehl commented on core/Tracker/Visitor.php:
```diff
@@ -42,10 +42,10 @@ public function setVisitorColumn($column, $value)
-    public function getVisitorColumn($column)
+    public function getVisitorColumn($column, bool $original = false)
     {
-        if (array_key_exists($column, $this->visitProperties->getProperties())) {
-            return $this->visitProperties->getProperty($column);
+        if (array_key_exists($column, $this->visitProperties->getProperties($original))) {
+            return $this->visitProperties->getProperty($column, $original);
         }
```
@bx80 could you please check this PR again. It still seems to contains some parts of the other PR that have been changed

bx80 replied:
Must have missed that in the rebase, I've reverted it now. :+1:

Commit:
Revert older immutable property changes, added back column creation for failing tests

Commit:
Revert older immutable property changes, added back column creation for failing tests

Commit:
Update system tests

Commit:
Update system tests

bx80 commented:
A few of the system test expected results for the `pageviews_before` field have changed with this PR which is a bit worrying as we've changed the calculation method but are expecting no change to the result.

I've dug into the failed tests to find out why and all the failing tests happen when the same non-url goal is converted multiple times in the same visit. Previously the `pagesviews_before` value was calculated at request time by linking conversions to actions, if there was visit that started with page view and then had multiple non-url conversions (e.g. custom events) then the old calculation would incorrectly multiply the single page view by each conversion, inflating the page views. 

So the previous calculation method was inaccurate for multiple non-url conversions in the same visit, this no longer happens now the value is calculated by using the visit `total_interactions` value. The system test results that failed were actually wrong and are being corrected by the updates here.

Commit:
Update system test

Commit:
Update system test

sgiehl commented on plugins/Goals/Columns/PageviewsBefore.php:
```diff
@@ -0,0 +1,64 @@
+                          $visitor->getImmutableVisitorColumn('visit_total_searches');
+
+        if ($action instanceof ActionPageview) {
+            $visitPageviews++; // The current action isn't yet included in visit total interactions
```
Is that one correct? Shouldn't the metric hold the number of page views that were done before?
Let's assume a goal is configured to be triggered automatically when a certain page is visited and a user directly visits that page. Should the metric then be 0 or 1? I would assume 0, as the user didn't visit any other pages before.

bx80 replied:
For attribution we want to divide the conversion between all the pages involved in the conversion, this would include the page that triggered the goal. So in this case with a goal triggered by directly accessing a page then the  metric should be 1. I suppose it's more pageviews_before_conversion than pageviews_before_viewing_this_page. This matches the specification in #2030 and the tests.

sgiehl commented on plugins/Goals/Commands/CalculateConversionPages.php:
```diff
@@ -0,0 +1,191 @@
+    {
+        $dates = $this->getInput()->getOption('dates');
+        $lastN = $this->getInput()->getOption('last-n');
+        $forceRecalc = ($this->getInput()->getOption('force-recalc') ?? 0);
```
You could also give that option a default value of `0`, then you wouldn't need to check that here

sgiehl commented on plugins/Goals/Commands/CalculateConversionPages.php:
```diff
@@ -0,0 +1,191 @@
+
+        foreach ($goals as $id) {
+            // validate the goal id
+            if (!$goalsModel->doesGoalExist($id, $idSite)) {
```
Not sure if the use case should be covered here, but this check would prevent passing a `0` for processing ecommerce orders.

bx80 replied:
Good spot, I've adjusted this to allow a `0` to be passed :+1:

sgiehl commented on plugins/Goals/PagesBeforeCalculator.php:
```diff
@@ -0,0 +1,154 @@
+            if ($idGoal === null) {
+                // All goals
+                $goalsModel = new GoalsModel();
+                $goals = array_column($goalsModel->getActiveGoals([$site]), 'idgoal');
```
Ecommerce orders shouldn't be processed?

bx80 replied:
Fixed :+1:

sgiehl commented on tests/PHPUnit/System/expected/test_CustomEvents__Actions.getPageUrls_month.xml:
```diff
@@ -97,11 +97,11 @@
-				<nb_conv_pages_before>3</nb_conv_pages_before>
-				<nb_conversions_attrib>0.9999</nb_conversions_attrib>
+				<nb_conv_pages_before>1</nb_conv_pages_before>
+				<nb_conversions_attrib>3</nb_conversions_attrib>
 				<nb_conversions_page_rate>1</nb_conversions_page_rate>
 				<nb_conversions_page_uniq>3</nb_conversions_page_uniq>
-				<revenue_attrib>10.33</revenue_attrib>
+				<revenue_attrib>31</revenue_attrib>
```
Any reason why all those test files changed their numbers "dramatically"? I would have actually expected that no test files would needed to be changed, as the new calculation should result in different numbers.

bx80 replied:
Please see the [comment](https://github.com/matomo-org/matomo/pull/20526#issuecomment-1539350546) I made above, the old calculation was inaccurate in some cases so a decrease in page views before is expected for some tests, the other metrics calculate from the pageviews_before so are also affected.

sgiehl commented on core/DataAccess/LogAggregator.php:
```diff
@@ -1107,80 +1107,42 @@ public function queryConversionsByDimension($dimensions = array(), $where = fals
+        $select = "
+            log_conversion.idvisit AS idvisit,
+            " . $idGoal . " AS idgoal,
+            " . ($linkField == 'idaction_url' ? '1' : '4') . " AS `type`,
```
does `1` and `4` here represent `Action::TYPE_PAGE_URL` and `Action::TYPE_PAGE_TITLE`. If so, please try to use the constants as that makes the code easier to understand

sgiehl commented on core/DataAccess/LogAggregator.php:
```diff
@@ -1107,80 +1107,42 @@ public function queryConversionsByDimension($dimensions = array(), $where = fals
+        $where .= sprintf('AND log_conversion.idgoal = %d
+                          AND logva.server_time <= log_conversion.server_time
+                          AND lac.type = %s',
                           (int) $idGoal, ($linkField == 'idaction_url' ? '1' : '4'));
```
same here for action types

sgiehl commented on plugins/Goals/PagesBeforeCalculator.php:
```diff
@@ -0,0 +1,154 @@
+                    $calcCount = $result->rowCount();
+
+                    // Done with this site/goal if no records were updated or we've processed 100m records (sanity check)
+                    if ($calcCount == 0 || $chunks > 10000) {
```
I don't understand the logic here. Why do you have the `while(!$done)` at all? The query actually updates all records it can find in the first run, as it doesn't have a certain limit that is changed with each run. So the first run updates everything and the second run always updates nothing and the while is aborted.
Also the check for `$chunks > 10000` looks unneeded. `$chunks` is reset for each site and increased with each run of the while loop. So it's increased by 2 for every goal that is processed. So 10.000 would be reached after the 5.000 goal was processed. While it might be possible that someone could have so many goals, I doubt it makes sense to abort then.

bx80 replied:
Sorry, this was confusing. This code originally included `LIMIT 10000` on the query so the updates were broken up into chunks, the while loop was to keep repeatedly running the update query until no more records needed updating. Each 'chunk' represents an update of 10,000 conversions, so the `$chunks > 10000` sanity check was to abort if there were more than 100,000,000 conversions. This worked well and had a callback method to update the UI after each chunk.

Unfortunately MySQL doesn't support update statements with joined tables and a limit clause so the limit clause was removed, this means the while / chunk logic no longer makes any sense so I've removed it :disappointed:

sgiehl reviewed (changes requested):
Left another couple of comments that came to my mind when looking through the code changes.

Commit:
Allow ecommerce calcs, use constants for action types, add parameter default, tidy loop logic

Commit:
Allow ecommerce calcs, use constants for action types, add parameter default, tidy loop logic

sgiehl commented on plugins/Goals/Commands/CalculateConversionPages.php:
```diff
@@ -180,7 +180,7 @@ private function getGoalsToCalculate(): ?string
         foreach ($goals as $id) {
             // validate the goal id
-            if (!$goalsModel->doesGoalExist($id, $idSite)) {
+            if (!$goalsModel->doesGoalExist($id, $idSite) && $id != 0) {
```
You could use `GoalManager::IDGOAL_ORDER` instead of `0`, so it easier to understand

sgiehl commented on plugins/Goals/PagesBeforeCalculator.php:
```diff
@@ -92,58 +91,49 @@ public function calculateFor(?string $startDatetime, ?string $endDatetime, ?int
                 // All goals
                 $goalsModel = new GoalsModel();
                 $goals = array_column($goalsModel->getActiveGoals([$site]), 'idgoal');
+                $goals[] = 0; // Include ecommerce orders
```
We should avoid processing the ecommerce goal if its not enabled.
```suggestion
                if (\Piwik\Site::isEcommerceEnabledFor($site) {
                    $goals[] = \Piwik\Tracker\GoalManager::IDGOAL_ORDER;
                }
```

sgiehl commented on core/Updates/5.0.0-b1.php:
```diff
@@ -110,4 +117,39 @@ private function requiresUpdatedLogVisitTableIndex()
+            return;
+        }
+
+        // Abort if there are more than 10,000 conversions across all sites and goals in the last 48hrs
+        $startDate = Date::factory('yesterday');
+        $endDate = Date::factory('today')->getEndOfDay();
+
+        $sql = 'SELECT COUNT(*) FROM ' . Common::prefixTable('log_conversion') . ' WHERE server_time <= ? AND server_time >= ?';
+
+        $result = Db::fetchOne($sql, [$startDate, $endDate]);
+        if ($result > 10000) {
+            return;
+        }
+
+        // Calculate all conversions for the last 48hrs
+        $pagesBeforeCalculator = new PagesBeforeCalculator();
+        $pagesBeforeCalculator->calculateFor($startDate, $endDate);
```
I wonder if we could convert this code to be a custom update instead. This would automatically suggest to run the command when updating manually on console. We did something similar for the token auth migration in Matomo 4. See https://github.com/matomo-org/matomo/blob/5.x-dev/core/Updates/4.0.0-b1.php#L62-L63

bx80 replied:
As discussed, I've removed the check to abort if query if there are 10,000 conversions. 

The migration has been updated to use a `CustomMigration` to run the `core:calculate-conversion-pages` command so that it is shown in the list of dry run commands when using `./console core:update`

Commit:
Use constant for ecommerce order goal id, only process if ecommerce enabled

Commit:
Use constant for ecommerce order goal id, only process if ecommerce enabled

sgiehl committed:
Merge branch '5.x-dev' into m20375-pages-before-field

Commit:
Remove update limit check

Commit:
Remove update limit check

Commit:
Use custom migration to call the CalculateConversionPages command during update

Commit:
Use custom migration to call the CalculateConversionPages command during update

Commit:
Merge remote-tracking branch 'origin/m20375-pages-before-field' into m20375-pages-before-field

Commit:
Merge remote-tracking branch 'origin/m20375-pages-before-field' into m20375-pages-before-field

Commit:
Fix PHPCS

sgiehl committed:
Merge branch '5.x-dev' into m20375-pages-before-field

sgiehl commented on plugins/Goals/PagesBeforeCalculator.php:
```diff
@@ -0,0 +1,154 @@
+
+                if (!empty($startDatetime)) {
+                    $sql .= " AND c.server_time >= ?";
+                    $bind[] = $startDatetime;
```
Did you consider timezones? When looking at this again I was wondering if that would work for sites that use another timezone than UTC. All data in the database is stored in UTC, but when a site is archived it might look at times that are before or after the UTC time. So guess we might need to apply something like this to the start and endtime:

```php
$timezone = Site::getTimezoneFor($idsite);
$bind[] = Date::factory($startDatetime, $timezone)->getDateTime();
```

bx80 replied:
Good point. The migration update for the switchover is for today and yesterday which is enough to cover any timezone differences, but standalone use of the command should adjust the timezone :+1:

Commit:
Apply site timezone to update command date parameters

Commit:
Merge remote-tracking branch 'origin/m20375-pages-before-field' into m20375-pages-before-field

sgiehl committed:
Merge branch '5.x-dev' into m20375-pages-before-field

sgiehl commented:
@bx80 The system tests on PHP 7.2 are currently running into a segfault. I'm able to reproduce that segfault on my vm.
It seems caused by any change in this PR, as the tests are passing on 5.x-dev. I'll check which test is causing it so we can further investigate...

sgiehl commented on core/Updates/5.0.0-b1.php:
```diff
@@ -48,6 +51,14 @@ public function getMigrations(Updater $updater)
+        $migrations[] = $this->migration->db->addColumns('log_conversion', ['pageviews_before' => "SMALLINT UNSIGNED DEFAULT NULL"]);
+
+        $instanceId = SettingsPiwik::getPiwikInstanceId();
+        if (strpos($instanceId, '.matomo.cloud') === false || strpos($instanceId, '.innocraft.cloud') === false) {
```
That is currently actually always `true`
```suggestion
        if (strpos($instanceId, '.matomo.cloud') === false && strpos($instanceId, '.innocraft.cloud') === false) {
```

bx80 replied:
:facepalm:

sgiehl commented:
The segfault occurs within the test BackwardsCompatibility1XTest. That test insets a database dump of Matomo 1.x and then tries to perform a database update here: https://github.com/matomo-org/matomo/blob/065efa10af6cc13f142ed157c1dcdf4f1b3f378b/tests/PHPUnit/System/BackwardsCompatibility1XTest.php#L44

This segfaults when it tries to perform the custom migration you have added. Commenting that part out lets the tests pass again. Don't have time right now to investigate that further...

bx80 committed:
Update core/Updates/5.0.0-b1.php

Co-authored-by: Stefan Giehl <stefan@matomo.org>

Commit:
Rework custom migration to return migration SQL object, fix segfault

Commit:
Directly query active goal ids for sites

bx80 commented:
@sgiehl I was able to recreate the segfault locally with PHP 7.4.33, strangely it doesn''t occur at all with with PHP 8.2.5.

It seems to be caused by the custom migration using the `plugins/Goals/Model/getActiveGoals()` method, directly retrieving the goal ids with a simple query doesn't produce the segfault. I can't see any obvious reason why using that method would cause the error.

As part of debugging this I've also reworked the custom migration to more closely match the 4.0.0 custom migration by generating an array of `BoundSql` objects and then using the `$updater->executeMigrations()` method instead of executing the queries directly.

sgiehl commented on plugins/Goals/Commands/CalculateConversionPages.php:
```diff
@@ -0,0 +1,339 @@
+     *
+     * @return array An array of queries and bind arrays   [['sql' => QUERY1, 'bind' => [PARAM1 => VALUE], ...]
+     */
+    public static function getQueries(?string $startDatetime, ?string $endDatetime, ?int $lastN = null, ?string $idSite = null,
```
Guess there is no need to have that method public. 
```suggestion
    private static function getQueries(?string $startDatetime, ?string $endDatetime, ?int $lastN = null, ?string $idSite = null,
```

sgiehl commented on core/Updates/5.0.0-b1.php:
```diff
@@ -48,6 +51,14 @@ public function getMigrations(Updater $updater)
         $migrations[] = $this->migration->db->addColumns('user_token_auth', ['post_only' => "TINYINT(2) UNSIGNED NOT NULL DEFAULT '0'"]);
+        $migrations[] = $this->migration->db->addColumns('log_conversion', ['pageviews_before' => "SMALLINT UNSIGNED DEFAULT NULL"]);
+
+        $instanceId = SettingsPiwik::getPiwikInstanceId();
+        if (strpos($instanceId, '.matomo.cloud') === false && strpos($instanceId, '.innocraft.cloud') === false) {
```
I'm still unsure if it's an appropriate way to exclude certain updates from running on cloud. 
Do we have an internal note, that makes sure that cloud team would run the command manually, so that won't be forgotten?
Just want to be super sure in this case, as we didn't do something like this before and I still think that if we expect problems on cloud that are reason enough to exclude it, we need to advice our high traffic on premise customers to skip that as well somehow.
@tsteur would you mind confirming this?

tsteur replied:
Good call @sgiehl . We could create an issue for the `CLOUD` project. We then put it in the 5.0 release and update our instructions to run this manually after the release. 
The code itself would otherwise work.

bx80 replied:
>if we expect problems on cloud that are reason enough to exclude it, we need to advise our high traffic on premise customers to skip that as well somehow.

The issue on cloud is running the update at the same time for thousands of customers when shared database resources are being used. Based on the query performance tests that have been done, large individual on-premise users shouldn't have a problem.

I've created issue CLOUD-1370 so this can be included in the Matomo 5.0 deployment tasks.

sgiehl commented on plugins/Goals/Commands/CalculateConversionPages.php:
```diff
@@ -0,0 +1,339 @@
+/**
+ * Command to calculate the pages viewed before conversions and populate the log_conversion.pages_before field
+ */
+class CalculateConversionPages extends ConsoleCommand
```
The code in this class is currently fully untested, right?
Was wondering if we maybe could add a simple test for that. Like tracking a couple of visits & conversions. Afterwards unsetting the `pageviews_before` column in the database and then running the command to check if the column is filled up correctly again...

bx80 replied:
I've added a new test `CalculateConversionPagesCommandTest`. It uses the same fixture as the `TrackGoalsPagesTest` which includes some of the more complicated conversion scenarios. Once the fixture is setup the test clears the `pageviews_before` column and then runs the command (both in date range mode and lastN) then compares the newly calculated `pageviews_before` values against the known values from the `TrackGoalsPagesTest`.

This test revealed an issue with the update query for visits which have multiple conversion of different goals, I've adjusted the query to calculate correctly for these conversions. The updated query is more complicated, so after optimizations I've retested it against a large dataset and it is updating 100,000 conversions with 4.5million linked visit actions in ~22 seconds which would still seem to be an acceptable level of performance.

sgiehl reviewed (changes requested):
Left some more comments. 
In relation to this PR we should maybe create a new FAQ to explain what to do if the page goal metrics aren't shown for older dates when e.g. a new segment is created. Running a command won't be something people might expect.
Not sure if that is something we might also need to mention somewhere in the release blog post.

bx80 committed:
Update plugins/Goals/Commands/CalculateConversionPages.php

Co-authored-by: Stefan Giehl <stefan@matomo.org>

Commit:
Add calculate conversion pages command tests, tweak update query to correctly handle multi-conversion visits

Commit:
Update query optimizations

sgiehl commented on plugins/Goals/tests/System/CalculateConversionPagesCommandTest.php:
```diff
@@ -0,0 +1,97 @@
+ * @group CalculateConversionPagesCommandTest
+ * @group Goals
+ * @group Plugins
+ * @group Plugins
```
```suggestion
```

sgiehl reviewed (approved):
Had a last look through the code and imho it looks fine now.
Added a last minor suggestions, which I will apply myself before merging it.

sgiehl committed:
remove duplicate test group
